### ISSUE TITLE
Batch size and optimizer 

### ISSUE THREAD

**Issue by nirandaperera** [msr-fiddle/pipedream#47]
Hi, 
I see that the profiler calculates memory and execution times based on a particular batch size. 
But the optimizer code does not take in any batch size parameter. So, does that mean, in the optimizer logic, the execution times and activation memories are normalized?

Best

**Comment by deepakn94** [comment]
Why do they need to be? Right now, the optimizer is not trying to sweep batch size? It assumes that this is provided as an input

**Comment by nirandaperera** [comment]
So the optimizer is equally dividing the work among all GPUs irrespective of the memory available?

**Comment by deepakn94** [comment]
No, that's not true. But why do you need to normalize by batch size to decide whether something fits? I am saying that the batch size is not a knob we try to tune -- given a batch size, we know the computation times and activation sizes, and we can use this information to make placement decisions for this particular batch size

**Comment by nirandaperera** [comment]
I see your point. So then if I am to run for different batch size, I have to start from the profiler, then optimizer,... etc? Just wanted to verify this. 

I was under the impression that the profiler can be reused for any batch size (of a particular model).

**Comment by deepakn94** [comment]
Right, that's correct. The computation time doesn't scale linearly with the batch size (throughput itself is a function of batch size), so you would probably want to run the profiler for each batch size anyway to get an accurate timing estimate (you could reuse activation size measurements, but we don't currently do this)...hope this clarifies!

**Comment by nirandaperera** [comment]
Thanks for the clarification.

**Comment by nirandaperera** [comment]
I have another question on the optimizer. I am running resnet101 on 4 Tesla K80's with 11GB of memory each. I am trying the straight pipeline. 
One thing I observed was, even if I run the optimizer with 9GB of memory, still the stage0 hangs on the bwd pass of the 0th batch. It seems to me that there is not enough memory in device0, but from cuda it silently hangs. 
I have attached a screenshot of the mem utilization. 
https://drive.google.com/file/d/1l_XRzGEmzEfQmNVFfBh96oXDPETk20in/view?usp=sharing

So, this would be a trivial question, but the optimizer takes into account the number of fwd activations a GPU needs to keep in memory (until its backward pass is later scheduled), isn't it?
